adr/ADR-19.md

```diff
@@ -0,0 +1,273 @@
+# JetStream based Object Stores
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-11-03|
+|Author  |@scottf|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client, objectstore|
+
+## Context
+
+This document describes a design of a JetStream backed object store.
+
+## Overview
+
+We intend to hit a basic initial feature set as below, with some future facing goals as indicated:
+
+Initial feature list:
+
+- Represent an object store.
+- Store a large quantity of related bytes in chunks as a single object.
+- Retrieve all the bytes from a single object
+- Store meta data regarding each object
+- Store multiple objects in a single store  
+- Ability to specify chunk size
+- Ability to delete an object
+- Ability to understand the state of the object store.
+
+Possible future features
+
+- Event Notifications (add/delete/lock)
+- Locking
+- Archiving (tiered storage)
+- Searching/Indexing (tagging)
+- Versioning / Revisions
+- Overriding digest algorithm 
+- Capturing Content-Type (mime type)
+- Per chunk Content-Encoding (i.e. gzip)
+
+## Naming Specification
+
+Protocol Naming Conventions are fully defined in [ADR-6](ADR-6.md)
+
+### Object Store
+The object store name or bucket name (`os-bucket-name`) will be used to formulate a stream name and is specified as: `restricted-term` or 1 or more of `A-Z, a-z, 0-9, dash, underscore`
+
+### Objects
+An individual object name is custom url encoding to form a portion of a subject (`os-object-name`).
+There are essentially no restrictions on the object name except that it has to be a valid filename across operating systems, so it is specified as: 
```

Comment:
It filename safe so it can be stored on any operating system. It gives a pretty wide latitude of what the name can be. Base64 would work as an encoding, but given the data in a filename that has to be encoded, url encoding only will regularly encode the dot, so it's much shorter.